packages/fondue/src/components/Tooltip/Tooltip.tsx

```diff
@@ -37,21 +37,22 @@ const formatTooltipText = (text: string) => {
 };
 
 type TimeoutRef = {
-    current: NodeJS.Timeout | null;
+    current?: number;
 };
-
 const handleTimeout = (callback: () => void, delay: number, timeoutRef: TimeoutRef) => {
```

Comment:
I know this is not part of your changes, but since we are refactoring this component, could we consider moving the `handleTimeout` within the component? 

This removes the need to create the `TimeoutRef` type and provides better cohesion with the side-effect operation performed by the function.

Reply:
I would like to make the main function readable and that is the main reason why it is out. The main ref `timeoutRef` is inside of the main function which prevents the side effects.

Reply:
Unfortunately, side-effects are not prevented since we are mutating the value of the `ref` internally.

If we want to improve readability, may I suggest we remove the need for that function altogether? It's only used in `handleShowTooltip` and `handleHideTooltip` and seems a premature abstraction.

Can I suggest the following refactor:

```tsx
   const handleShowTooltip = () => {
        if (disabled) {
            return;
        }

        if (timeoutRef.current) {
            window.clearTimeout(timeoutRef.current);
        }

        if (enterDelay === 0) {
            setOpen(true);
        } else {
            timeoutRef.current = window.setTimeout(() => {
                setOpen(true);
            }, enterDelay);
        }
    };

    const handleHideTooltip = useCallback(() => {
        if (timeoutRef.current) {
            window.clearTimeout(timeoutRef.current);
        }

        if (leaveDelay === 0) {
            setOpen(false);
        } else {
            timeoutRef.current = window.setTimeout(() => {
                setOpen(false);
            }, leaveDelay);
        }
    }, [leaveDelay]);
```
Yes, there is a little bit of duplication (3 lines?) but we remove altogether the need to:

- have a shared function
- pass the ref within it
- pass a callback

What do you think?

Reply:
Unfortunately, this creates more cognitive load on the simple function and main function, and the the handle methods have to much responsibility and it is not clear where all the flow happens inside of the main function. and also the function should to less things and not more to easier grasp the content.

I can not relate to that because the var is coming from the main component.

Reply:
Yes, I think I can agreed on the multiple responsibility (state update + timer management). 

I would like to propose an alternative solution to avoid side-effects outside of the component:

```ts
    const delayedSetOpen = (show: boolean, delay: number) => {
        if (timeoutRef.current) {
            window.clearTimeout(timeoutRef.current);
        }

        timeoutRef.current = undefined;

        if (delay === 0) {
            setOpen(show);
        } else {
            timeoutRef.current = window.setTimeout(() => {
                setOpen(show);
            }, delay);
        }
    };

    const handleShowTooltip = () => {
        if (disabled) {
            return;
        }

        delayedSetOpen(true, enterDelay);
    };

    const handleHideTooltip = () => {
        delayedSetOpen(false, leaveDelay);
    );

```

Reply:
I appreciate your input, its helpful and would like to propose another solution to avoid side effects:
```
this function is still outside of the main component 
const handleTimeout = (callback: () => void, delay: number, timeoutRef: RefObject<number | undefined>) => {
    let id = timeoutRef.current;

    if (id) {
        window.clearTimeout(id);
    }

    if (delay > 0) {
        id = window.setTimeout(callback, delay);
    } else {
        callback();
    }

    return id;
};

and this are inside the Tooltip method
const handleShowTooltip = () => {
        if (!disabled) {
            timeoutRef.current = handleTimeout(() => setOpen(true), enterDelay, timeoutRef);
        }
    };

    const handleHideTooltip = useCallback(() => {
        timeoutRef.current = handleTimeout(() => setOpen(false), leaveDelay, timeoutRef);
    }, [leaveDelay]);
```

If the name is not right in your opinion I would not name it `delayedSetOpen` because the funciton is also used for delayed closing the Tooltip.

Reply:
Yes, I like your proposal. If we have a signature of:

```ts
type HandleTimeout = (cb: () => void, delay: number, timer?: number): number 
```
we have removed the side-effect from its body.

```ts
const handleTimeout = (callback: () => void, delay: number, timer?: number) => {
   let newTimer: number;

    if (timer) {
        window.clearTimeout(timer);
    }

    if (delay > 0) {
        newTimer = window.setTimeout(callback, delay);
    } else {
        callback();
    }

    return newTimer;
};
```

With this change, since we want to keep it outside from the component, we have completely decoupled the logic from React APIs.

Reply:
yes true, to decouple and get rid of side-effects.
I just left null because it is the clear timeout is using it.
```
const handleTimeout = (callback: () => void, delay: number, timeoutId: number | null) => {
    let newTimeoutId = timeoutId;

    if (newTimeoutId) {
        window.clearTimeout(newTimeoutId);
    }

    if (delay > 0) {
        newTimeoutId = window.setTimeout(callback, delay);
    } else {
        callback();
    }

    return newTimeoutId;
};
```

Reply:
If we want pure functions in this case :)